File: 032/main.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function (s) {
  let ans = 0
  const vis = []
  const stack = []
  for (let i = 0; i < s.length; i++) {
    vis.push(0)
    const item = s[i]
    // number less than 0 represents '('
    // number greater than or equal to 0 represents the index of the previous '('
    if (item === ')' && stack.length && stack[stack.length - 1] >= 0) {
      const left = stack.pop()
      if (left === 0) {
        vis[i] = i - left + 1
      } else {
        vis[i] = vis[left - 1] + i - left + 1
      }
      ans = Math.max(ans, vis[i])
    } else {
      stack.push(item === '(' ? i : -1)
    }
  }
  return ans
}

if (process.env.LZS) {
  // local test
  const assert = require('chai').assert
  const Chance = require('chance')
  const chance = new Chance()
  const brute = function (s) {
    let maxx = 0
    const valid = function (s) {
      const stack = []
      for (const item of s) {
        if (item === '(') {
          stack.push(item)
        } else { // ')'
          if (stack.length === 0 || stack[stack.length - 1] !== '(') {
            return false
          }
          stack.pop()
        }
      }
      return stack.length === 0
    }
    for (let i = 0; i < s.length; i++) {
      for (let j = i + 1; j < s.length; j++) {
        let substr = s.slice(i, j + 1)
        if (valid(substr)) {
          maxx = Math.max(maxx, substr.length)
        }
      }
    }
    return maxx
  }
  assert.strictEqual(longestValidParentheses(''), 0, 'empty string')
  assert.strictEqual(longestValidParentheses('(('), 0, '((')
  assert.strictEqual(longestValidParentheses('(()'), 2, '(()')
  assert.strictEqual(longestValidParentheses('(())'), 4, '(())')
  for (let _ = 0; _ < 100; _++) {
    const len = chance.integer({ min: 1, max: 200 })
    const s = Array.from({
      length: len
    }, () => chance.character({ pool: '()' })).join('')
    assert.strictEqual(longestValidParentheses(s), brute(s), s)
  }
}

File: 032/main.py
class Solution(object):

    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack, leftmost = [], -1
        maxx = 0
        for i, item in enumerate(s):
            if item == '(':
                stack.append(i)
            else:
                if not stack:  # empty
                    leftmost = i
                else:
                    if stack:  # not empty
                        stack.pop()
                    if stack:
                        maxx = max(maxx, i - stack[-1])
                    else:
                        maxx = max(maxx, i - leftmost)
        return maxx

import os
import random
import unittest

if os.getenv('LZS'):
    def valid(s):
        stack = []
        for item in s:
            if item == '(':
                stack.append(item)
            else:
                if not stack:
                    return False
                stack.pop()
        return not stack

    def brute(s):
        maxx = 0
        for i in range(len(s)):
            for j in range(i + 1, len(s)):
                if valid(s[i:j + 1]):
                    maxx = max(maxx, j - i + 1)
        return maxx

    class Test(unittest.TestCase):

        def test_solution(self):
            for _ in range(100):
                length = random.randint(10, 100)
                s = Solution()
                parentnesses = [random.choice('()') for _ in range(length)]
                res, expect = s.longestValidParentheses(
                    parentnesses), brute(parentnesses)
                self.assertEqual(
                    res,
                    expect,
                    '{} != {} for {}'.format(res, expect, parentnesses)
                )
    unittest.main()
